rfd/0149-headless-kube.md

```diff
@@ -0,0 +1,257 @@
+local proxy ([RFD 121](https://github.com/gravitational/teleport/blob/master/rfd/0121-kube-mfa-sessions.md)).
+
+## Headless authentication
+
+When headless authentication is initiated by the user on a remote machine, the client calls the `POST /webapi/login/headless`
+endpoint on the Auth server. It creates a `HeadlessAuthentication` resource on the backend.
+After that, the call is paused, and the user is provided with a link (or command line alternative) where they can approve
+the newly created headless authentication request using their local machine with a logged-in Teleport and an MFA device. Once
+the request is approved by the user, the authentication process is unpaused and the original call returns regular
+user's Teleport login certificates created by the auth server. These certificates have a TTL of 1 minute.
+
+## Headless Kubernetes proxy
+
+We will add the flag `--headless` to the `tsh proxy kube` command.
+The user will be able to start `tsh --headless proxy kube` in headless mode on a remote machine,
+then authenticate on the local machine using the link provided by the started proxy (using the flow
+described in the previous section).
+
+After the user finishes the authentication process, the headless kube proxy will generate certificates for accessing
+specified Kubernetes clusters. When the Auth server issues these certificates, it will recognize that
+the certificate is being generated for the user kube proxy - certificate usage is `UsageKubeOnly` and
+`RequesterName` is `TSH_KUBE_LOCAL_PROXY`. For this case, it will not limit the certificate TTL by the calling
+identity TTL, which is 1 minute, since that's what was created during headless login. Instead, these certificates'
+expiration will be limited by the `max_session_ttl` time, similar to how the regular `tsh proxy kube` works.
+When the kube proxy receives certificates for accessing Kubernetes clusters, it will create an in-memory file and
+will write kubeconfig to it. This kubeconfig is only needed for correct access from the kubectl to the local kube proxy;
+it does not contain certificates information for accessing the Kubernetes cluster in teleport. To create the kubeconfig we
+generate a new one-use local key every time the proxy starts. Next, the headless kube proxy will reexec itself, opening a new
+shell for the user, while keep running in the background. The kube proxy will provide this new shell with the
+access to the in-memory kubeconfig by the path `/proc/self/fd/3` and the proxy will point the `$KUBECONFIG`
+environment variable to this file, allowing `kubectl` to access it.
+
+On MacOS, we will write kubeconfig to a regular temporary file, restricted by the permissions to be available
+only to the logged-in user.
+
+After the kube proxy reexecs into a new shell, the user is ready to use `kubectl`. Even though the main user certificate received during
```

Comment:
I wonder if we should support some kind of idle timeout to mitigate the fact that you can basically leave your remote session open for 8+ hours with valid credentials. Maybe as a future improvement.

Reply:
Yep, I guess we can add that in the future. Maybe it can be more wider thing for `tsh proxy` commands.